### ISSUE TITLE
[css-borders] Add a 'hairline' border-width value

### ISSUE THREAD

**Issue by tabatkins** [w3c/csswg-drafts#3720]
People regularly ask for the ability to size things in device pixels (#3715 is the latest). This is virtually *always* intended for the use-case of sizing borders, to create "hairline" borders or separators.  

Since adding an actual device pixel unit is out of the question, we should still address this use-case directly, and add a `hairline` keyword to 'border-width', representing "the thinnest width that the UA believes is recognizable, snapped to the nearest whole number of device pixels". *Right now*, that'll be 1 device pixel, but if screens get significantly higher-res, it'll probably resolve to whatever number of device pixels gives you a value between 1/3 and 1/2 of a CSS px.

--------

## Why can't we add a device pixel unit, so you could just say `border-width: 1dp`?

Because device pixels are unpredictable in size. If you try to size anything bigger than a hairline with device pixels, the size will, today, vary by more than 300%, as there are devices that have 1dp==1px, and devices with 3dp=1px, and I think some devices are pushing higher than that. As display technology gets better,  this gulf will continue to grow, as there will likely *always* be 1dp==1px devices in use.

We already know from past experience that authors can't generally handle units with an unpredictable "sliding scale" ratio; the CSS2 spec used to have `in` and `px` decoupled, so that `in` was as close to an actual inch as the browser could get, and `px` was the nearest whole number of device pixels approximating the visual angle of a device pixel on a 96dpi monitor. However, so many authors used the physical units (`pt`/`in`/`cm`/`mm`) to size their pages while assuming 96px=1in that browsers implementing the spec got broken pages on devices whose in-px ratio was different. And that was a ratio that could differ whose min and max could differ by, at most, 50%!  (Handling an unpredictable ratio is a fundamentally very difficult problem! We have better, less fragile layout tools to help with that these days, but it's still virtually guaranteed to result in problems if we expose this.)

Plus, the notion of a "device pixel" is screen-specific; when printing, the analogue to "device pixel" is the "dot", the smallest drop of ink that can be placed at a particular location. Printers not only have much *smaller* dots than screens have pixels, and in a larger range (consumer printers can vary from 300dpi to more than 1000dpi), they also vary based on color; black-and-white dpi is often much higher than color dpi. Sizing *anything* based on "device pixels" will either be *wildly unpredictable* on printers, or else it'll be a "virtual device pixel" that doesn't correspond to anything physical anyway.

And all this isn't even getting into screens that simply more exotic than the "rectilinear grid of hardware pixels" design that this sort of feature assumes. Some screens layer their pixels in a non-square grid. Some have *differently-sized* pixels, layered in separate rectilinear grids. Some have different grids for each subpixel color.  All in all, the notion of a "device pixel" as something you can naively size things in is already nonsensical for many output devices, and the situation is likely to only get worse as time goes on.

## Why not a device-pixel rounding function?

In #3715 @florian and I talk about a function that rounds a length to the nearest number of device pixels, as a compromise that delivers many of the benefits of device-pixel sizing without as many of the downsides. On further consideration, I don't think we want this either.

For the case of hairline borders, a device-rounding function achieves basically the same thing as a 'hairline' value, just in a slightly more complicated way. Users have to know to write `border-width: dp-round(.5px)` or what-have-you, rather than `border-width: hairline;`.  The only theoretical benefit of `dp-round()` over `hairline` is that it lets you distinguish between whether you *really* want .33px or .5px, when that distinction might matter for visibility, but in practice that's a wash anyway, since a 3x screen will still probably choose .33px over .67px if you say `dp-round(.5px)`, and a 2x screen will still use .5px if you say `dp-round(.33px)`, and many devices will end up still choosing 1px anyway, so the theoretical gain in precision is lost in practice. In a theoretical future with super high-res screens, it might become relevant; we can revisit the question at that point. ^_^

For *any other case*, rounding to the device pixel doesn't do anything meaningful *unless you're also rounding the position*. Otherwise it's possible that the box is exactly 200dp tall, but it's offset 150.2dp from the top of the screen, so it's still fuzzy. And unless we offer position-rounding as some sort of separate property switch, you have to instead use dp-round() on your *ancestors*, too, all the way up to the root. (And depending on how you're being laid out, that still might not be enough; flexbox alignment has no way to force a rounding on the gap between flex items, for example.)

So, dp-round() is overkill complexity for hairlines, and doesn't do anything useful for most non-hairline cases.

**Comment by AmeliaBR** [comment]
I think this is a good compromise to the issue, especially if there is spec text recommending that browsers consider accessibility when defining the "hairline" thickness.  E.g., if I'm browsing the web with 400% zoom level, I probably don't want lines that are only one retina-screen-pixel thick!

For the more general case of "snapping" lines to an even number of device pixels, the [SVG `shape-rendering` property](https://svgwg.org/svg2-draft/painting.html#ShapeRendering) is another approach, which could be generalized to apply to CSS borders and boxes.  However, `shape-rendering: crispEdges` wouldn't remove the need for a dedicated `hairline` option, since snapping very fine lines to the nearest number of device pixels [sometimes results in snapping them down to nothing](https://mobile.twitter.com/AmeliasBrain/status/1087407083522146304)!

**Comment by tabatkins** [comment]
Yes, the spec should definitely state that the hairline-ness should be taken at "normal" zoom levels; zooming in shouldn't *keep* them hairline!

**Comment by dbaron** [comment]
It seems like at some point these borders might become thinner than the user of this keyword expects.  For example, when printing, or as devices get even-higher-resolution.

Also... with Gecko's border rounding behavior (round down in device pixels, except round values between 0 and 1 device pixels up to 1) you can effectively do this today without facing that problem.  For example, a border with width `0.25px' will be one device pixel, unless there are 8 or more device pixels per CSS pixel.  I think there was some movement of other browsers towards that behavior, but I'm not sure if things have converged.

**Comment by tabatkins** [comment]
@dbaron Did you miss this part (quoted below), or is your comment asserting that something about it would still be problematic?

> Right now, that'll be 1 device pixel, but if screens get significantly higher-res, it'll probably resolve to whatever number of device pixels gives you a value between 1/3 and 1/2 of a CSS px.

--------------

> Also... with Gecko's border rounding behavior [snip]

Hm, interesting. <http://software.hixie.ch/utilities/js/live-dom-viewer/saved/6749> tests this behavior. In Chrome, on my 1x desktop monitor, I get a 1px border for everything down to .05px; the .01px border disappears. That might be an internal rounding issue?

There's an additional weirdness in Chrome, tho - despite the border displaying the same size in all of them, the *internals* of the box itself lays out as different heights, despite the only thing giving it a height being a `padding: 2px` declaration! Looks like the 1px-border is 4px tall internally, the .5px and .33px are 3px tall, and the .1px is 2px tall. It looks like we're rendering the border-box's height "correctly" for the specified padding+borders (6px, ~5px, and ~4px, respectively), but then just *painting* the borders larger, so it overlaps the padding box internals?

It looks like Firefox correctly renders a 4px padding box on each of them, so I assume this is a Chrome bug in the timing of our border-width snapping vs other layout calculations.

**Comment by tabatkins** [comment]
(Confirmed with @atotic, this is a Chrome bug that'll be fixed when they get around to rewriting the relevant chunk of our layout code into the new LayoutNG form.)

**Comment by dbaron** [comment]
I did miss that "Right now" part -- although I think the proposal needs to define that behavior today, for printing.

**Comment by jonjohnjohnson** [comment]
When I don't want a keyline to effect geometry, I've used `box-shadow`. I don't s'pose we'd also be able to solve this `hairline` value for properties like `box-shadow` & `outline-width`, let alone hopeful properties like `text-stroke-width`([compat](https://compat.spec.whatwg.org/#the-webkit-text-stroke-width))?

**Comment by tabatkins** [comment]
> I did miss that "Right now" part -- although I think the proposal needs to define that behavior today, for printing.

Right, sorry for not being clear, it is *absolutely* intended that the spec will *explicitly* say that UAs should choose an appropriate small width, and recommend that it probably be no smaller than 1/3 or 1/2 of a px.

**Comment by tabatkins** [comment]
> I don't s'pose we'd also be able to solve this hairline value for properties like box-shadow & outline-width, let alone hopeful properties like text-stroke-width(compat) or text-shadow?

Hm, maybe.

Tho at this point dbaron's statement about just relying on rounding behavior (and specifying that you must *always* round away from zero) is starting to sound *really good*.

I had a chat with @bfgeek about this tho, and he has some reservations about it. He could elaborate on this more; I don't fully understand the problems he was alluding to.

**Comment by cncuckoo** [comment]
I do think the supporting for fractional order `px` value, means 0.1~0.9px (may be even more precise than this), is more useful than we finally only have a new `hairline` key word value for `border-width` property.

As on mobile web, Android app can use `0.3dp` or `0.5dp` , and iOS app can use `0.3pt` or `0.5pt` to express the hairline line. For CSS, `px` just play the same role as `dp` on Android and `pt` on iOS.

I wish, one day, we can use `0.3px` or `0.5px` to match the exact border width with the mobile apps, so the users can see the same thick or thin lines on the their Apps.

**Comment by AmeliaBR** [comment]
> I don't s'pose we'd also be able to solve this hairline value for properties like box-shadow & outline-width, let alone hopeful properties like text-stroke-width(compat) or text-shadow?

The `outline-width` property references the [`<line-width>` compound type](https://drafts.csswg.org/css-ui-4/#propdef-outline-width) that is defined for `border-width` and therefore accepts border-width keywords.  So it should automatically also get a `hairline` value if it is added as a new keyword.

`stroke-width` as defined in SVG doesn't accept keyword values, but there's no reason we couldn't add them (meaning, `thick`, `medium`, `thin`, as well as `hairline`)  for [the new module that extends it to CSS text](https://drafts.fxtf.org/fill-stroke/#stroke-width).

`box-shadow` and `text-shadow` would be more complicated, since there are many different lengths (offset and expansion/inset) interacting.  But maybe it's worth a separate issue to discuss the expected behavior of fine resolution shadow effects — I recently discovered that browsers are very inconsistent in how they handle `text-shadow` offsets of less than 1px (see [Twitter thread with people posting screenshots](https://mobile.twitter.com/AmeliasBrain/status/1105258703878418432)).

**Comment by AmeliaBR** [comment]
> I do think the supporting for fractional order px value … is more useful than we finally only can have a new hairline key word value

There's nothing stopping you from using a fractional px value for border width in CSS — it's just not guaranteed to snap to an even number of device pixels, and will therefore look bad on a screen with a different resolution than you designed it for.  If this is really important, however, you could use [media queries to check for the screen resolution](https://drafts.csswg.org/mediaqueries/#resolution).

The benefit of the keyword is that the browser would figure out the exact value required to get a crisp line.

**Comment by bradkemper** [comment]
> There's nothing stopping you from using a fractional px value for border width in CSS — it's just not guaranteed to snap to an even number of device pixels

I think it should. If it was only a 1x screen, then use opacity to simulate the thinner line, like is typically done for antialiasing scaled images. So a 0.5px line is rendered with 1px at half the opacity of what it normally would be. On a 3x screen, it would use 2 device pixels, at 0.75 opacity.

**Comment by tabatkins** [comment]
> There's nothing stopping you from using a fractional px value for border width in CSS — it's just not guaranteed to snap to an even number of device pixels,

It actually is guaranteed to do that. CSS itself doesn't guarantee it, but all browsers snap borders to whole device pixels, precisely *because* they otherwise look bad. (And do some fancy rounding to make sure all borders of a given width round in the same direction.)

**Comment by bradkemper** [comment]
@tabatkins I thought Chrome and Safari (and maybe others) snapped to whole px, not to device pixel smaller than px. Is that no longer true?

**Comment by bradkemper** [comment]
At least for border widths, I mean. Things like centering might be on device pixel space.

**Comment by tabatkins** [comment]
No, they definitely do device pixel snapping. Confirmed by looking at <http://software.hixie.ch/utilities/js/live-dom-viewer/saved/6749> on my high-dpi phone; the first box (`border-width: 1px`) definitely has wider borders than all the others.

**Comment by bradkemper** [comment]
Hmm. You're right. But they all (except 1px as and 0px) seem to be rounded to the same value on my iPhone X. Looks like maybe 0.25px. Maybe we can standardize and improve this then? It seems better than 'hairline' to me. But maybe both are useful.

**Comment by bradkemper** [comment]
On my retina desktop display (Mac), in Chrome for Mac, the 0.05px to 1px borders all look the same (1px), and the last 2 are rounded down to zero.

**Comment by cncuckoo** [comment]
>  If this is really important, however, you could use media queries to check for the screen resolution.

@AmeliaBR Yes, I could. But when I viewed http://software.hixie.ch/utilities/js/live-dom-viewer/?saved=6749 on Chrome/Safari/Firefox/Opera Mini on my iPhone XS Max, as @bradkemper has pointed, all values that less than 1px are rounded to 0.25/0.3px (maybe 1 device pixel). Based on this, I think using media queries to check the screen resolution is also not useful. Because it seems 
 none of the browsers have yield the control to us for how many device pixel one fractional px value could be snapped to.

**Comment by SelenIT** [comment]
It seems that different properties are snapped to device pixels [by different rules](https://codepen.io/SelenIT/pen/LbbGZL). Can anybody shed some light on the general logic behind this, please?

**Comment by FremyCompany** [comment]
My two cents here, maybe we should instead have a `border-style: hairline` and `outline-style: hairline` that would draw a perfect border around the content inside, and potentially leave extra blank space when that size is smaller than the border-width.

That way, the layout doesn't become unstable when you zoom in, only the rendering can change.

**Comment by FremyCompany** [comment]
I'm personally mostly concerned about layout changing as you zoom, this could have unintended consequences.

**Comment by bradkemper** [comment]
@fremycompany zooming should still scale it, along with everything else.

**Comment by FremyCompany** [comment]
@bradkemper Wait, so when you zoom, the "hairline" stops being an hairline, did I get that right?

**Comment by bradkemper** [comment]
@fremycompany 

Yes, Tab replied to Amelia’s comment with this:

> Yes, the spec should definitely state that the hairline-ness should be taken at "normal" zoom levels; zooming in shouldn't keep them hairline!

I take zooming to be like looking through a magnifying glass, where a hair (or hairline) looks bigger and thicker as a result.

**Comment by bradkemper** [comment]
> Because for instance if you had specified 1hairline, landed on 1dpx, then zoom 1.5x then the new border size won't be a round number of pixels (1.5dpx in this case) and won't be an hairline. I guess that this is an option, but that doesn't sound great to me.

That’s no worse than zooming in 1.5x on 1px line when 1dpx = 1px (non-retina display), is it?

**Comment by FremyCompany** [comment]
> > Because for instance if you had specified 1hairline, landed on 1dpx, then zoom 1.5x then the new border size won't be a round number of pixels (1.5dpx in this case) and won't be an hairline. I guess that this is an option, but that doesn't sound great to me.
> 
> That’s no worse than zooming in 1.5x on 1px line when 1dpx = 1px (non-retina display), is it?

Fair point.

**Comment by faceless2** [comment]
For what it's worth, the hairline concept will translate very easily when printing. A line width of 0 in PDF means "as thin as the output device can manage". When viewing the PDF on screen they're expected to remain as thin as possible regardless of zoom level.

**Comment by 1jj** [comment]
I think this can easily be done with fractional pixels in all browsers these days, right?
At least for me `border-width: 0.6px` or `0.5px` does it pretty well.

If it's not specified, it should be. Probably what @dbaron dbaron said above plus a lower limit of 0.3 CSS pixels or whatever.

> Also... with Gecko's border rounding behavior (round down in device pixels, except round values between 0 and 1 device pixels up to 1) you can effectively do this today without facing that problem. For example, a border with width `0.25px' will be one device pixel, unless there are 8 or more device pixels per CSS pixel. I think there was some movement of other browsers towards that behavior, but I'm not sure if things have converged.

**Comment by Loirooriol** [comment]
See https://drafts.csswg.org/css-values-4/#snap-a-length-as-a-border-width (#5210)

0.5px will result in 1 device pixel when the resolution is less than 4dppx, 2 device pixels between 4ddpx and 6dppx, etc.

If you want 1 device pixel regardless of the resolution, you can go smaller like `0.01px`, but then you risk that it could become 0px due to precision limitations before being snapped as a border width. For example, Firefox stores it as an app unit with a precision of 1/60 pixels, thus you get no border for `0.0083333333px`. Blink and WebKit store the length as a floating-point number, so you can go much lower.

**Comment by PsiRadish** [comment]
This sounds great as long as there's also a new unit called "hair" I can use inside `calc` to easily match sizes between things with and without borders. Something like this:
```css
.thing-with-border
{
    width: 32px;
    border: hairline solid magenta;
}
.thing-without-border
{
    width: calc(32px + 2hair);
}
```

**Comment by emilio** [comment]
So right now you can get `hairline` border widths if you have `border-width: 0.01px` (or so), where the engine will round up to 1 device pixel.

In https://github.com/w3c/csswg-drafts/issues/10819 I presented a use-case that isn't quite covered by this, because I actually want to use it to something like a shadow. I think the solution proposed there (previously proposed independently in https://github.com/w3c/csswg-drafts/issues/5986, cc @Crissov) of just a single environment variable is a good compromise between not making this too easy to use, yet still providing reasonable values for when you do need this.

Thoughts @tabatkins?

**Comment by tabatkins** [comment]
Do we want to expose the resolution, or just expose the width of a hairline? Are there enough good uses for the resolution that we actually want it, versus just adding a `hairline` keyword to a few properties?

**Comment by tabatkins** [comment]
Sorry, hit submit too early.

I'm still loathe to expose resolution directly, because a hairline isn't necessarily 1 device pixel. In a hypothetical super-display with 100 device pixels per 1px length, a hairline definitely shouldn't be .01px.

(And this also argues against just relying on border-width rounding away from zero to the nearest device pixel, in favor of a `hairline` keyword that exposes the "correct" width regardless of resolution.)

**Comment by yisibl** [comment]
So how do you solve the problem of displaying borders larger than `1px`?

**Comment by emilio** [comment]
> Do we want to expose the resolution, or just expose the width of a hairline? Are there enough good uses for the resolution that we actually want it, versus just adding a hairline keyword to a few properties?

I think exposing the resolution is the least implementation complexity, and the most flexible, since it's property agnostic... E.g., I might want to use padding in order to reserve space for a box-shadow `hairline`... Or I could imagine wanting an nth multiple of hairlines.

> (And this also argues against just relying on border-width rounding away from zero to the nearest device pixel, in favor of a hairline keyword that exposes the "correct" width regardless of resolution.)

Sure, though this behavior is unlikely to be changeable? It's fairly long-standing in some way or another...

**Comment by frivoal** [comment]
> > In a hypothetical super-display with 100 device pixels per 1px length, a hairline definitely shouldn't be .01px
> 
> What should it be?

Something in the 0.3px to 0.5px range probably. Maybe a bit lower, but not a lot lower. 0.01px, on a device that has pixels small enough to do it accurately, would be too thin to be visible.

**Comment by frivoal** [comment]
Also, while the layout size of a hairline probably needs to be fixed to avoid weird unstable layouts, the painted line within that size should probably be rounded / snapped to an integer number of pixels, taking position into account, in order to avoid blurriness. Although, if the physical pixels are small enough, that may not even be necessary.

**Comment by tabatkins** [comment]
> I might want to use padding in order to reserve space for a box-shadow hairline... 

Sure, we could allow hairlines there.

> Or I could imagine wanting an nth multiple of hairlines.

Sure, we could even make hairline just an absolute length unit, like `2hairline`, or it can be a calc() keyword that resolves to a length, like `calc(2 * hairline)`. Either route makes hairlines usable *everywhere*, and avoids any potential grammar issues that could arise from adding a keyword to the `<length>` type.

> Sure, though this behavior is unlikely to be changeable? It's fairly long-standing in some way or another...

Yeah, not saying we should get rid of the behavior, just that we might not want to *rely* on it for the hairlining, for several reasons. (And we might want to clamp it to the actual hairline width, rather than "nearest device pixel", so future screens don't round up to a nearly-invisible width.)

> What should [the size of a hairline on a screen with 100 device pixels per px] be?

Like Florian said, around half or a third of a px, probably. That's a question for designers and a11y folks - whatever the minimal visible width of a "hairline" that's appropriate. My point is that a "hairline border" is a meaningful concept that has *no* connection to device pixels, except insofar as the device pixels on many screens are currently approximately the right width for a hairline.

> But it's not just a matter of knowing the resolution, in order to avoid blurriness authors would also need the ability to align things correctly with the device pixels?

Right, rounding the width to device pixels is important for usability, so just exposing a resolution directly doesn't actually solve the problem, afaict.

> the painted line within that size should probably be rounded / snapped to an integer number of pixels, taking position into account, in order to avoid blurriness

Border snapping already handles that for us, luckily.

**Comment by kbabbitt** [comment]
> the layout size of a hairline probably needs to be fixed to avoid weird unstable layouts

This was my only concern as I was catching up on this discussion. Can we specify as part of the definition of `hairline` that, for sizing purposes, it's treated as equivalent to `1px`?

Though that might cause counterintuitive results if we also allow for multiples of `hairline`. Suppose `3 * hairline` resolves to fewer than 1px worth of device pixels. Does the resulting border occupy 3px of layout space even though the line is painted as only 1px wide?

**Comment by Loirooriol** [comment]
> Can we specify as part of the definition of hairline that, for sizing purposes, it's treated as equivalent to 1px?

Do you mean that with 2dppx, a hairline border would be sized as 1px but then only 0.5px would be painted and the other 0.5px would be transparent?

I don't think that authors will like getting these gaps when trying to line up elements correctly.

**Comment by kbabbitt** [comment]
I guess my question is, how do we ensure predictable layout if we don't define the space taken up by a hairline border? Suppose we don't do that, and an author using a 2dppx display writes:

```css
.container {
  width: 500px;
}
.tile {
  display: inline-block;
  width: 99px;
  height: 99px;
  border: hairline solid black;
}

<div class="container">
  <div class="tile"></div>
  <div class="tile"></div>
  <div class="tile"></div>
  <div class="tile"></div>
  <div class="tile"></div>
</div>
```

On their 2dppx display, the border-box width of each tile is 100px, and the five tiles fit exactly within the container. But for a user on a 1dppx display, the border-box width of each tile is 101px, and the last tile wraps to a new line, thus breaking the intended layout. That seems worse to me than having small transparent gaps between the tiles on a high-res display.

**Comment by emilio** [comment]
That'd be how they work, and what you're describing right now is how browsers behave if you `s/hairline/0.01px` in your test-case fwiw. Exposing the resolution directly would make it clearer IMO, i.e., you'd do `calc(1px / env(resolution))` and it's easy to see why at different resolutions you get overflow vs. not...

**Comment by tabatkins** [comment]
Chatted with Kevin in person, and I appreciate their idea more fully now.

The idea is that a hairline border is 1px wide, geometrically, but we *paint* it using smaller device pixels if possible (UA-defined). The device pixels are chosen to be an "inner" border, so the padding box remains the specified size and we can make the outer device pixels transparent. (Possibly we could offer a choice here, but we'd need to define what to paint on the inner device pixels if it's painted into the "outer" ones. Probably not needed; I'd definitely say avoid it until it's asked for with a good example.)

This should definitely apply in the other border-like properties, like column-rule, but there we should define it as choosing a *centered* set of device pixels (as centered as possible, at least).

The benefit of this is that the layout size of a hairline border is predictable across devices. We already know from the historical example of trying to allow the physical lengths to be true (with a varying ratio with `px`) that authors very commonly will design a layout that works on their device, but breaks on one with a different unit ratio. So, having a hairline border *actually* be sized to its visible width has real potential to cause accidentally broken layouts. It also would require a relayout when moving a window between screens with potentially different resolutions (but maybe we do that already?), while Kevin's idea just requires a repaint.

> Exposing the resolution directly would make it clearer IMO, i.e., you'd do calc(1px / env(resolution)) and it's easy to see why at different resolutions you get overflow vs. not...

My argument that hairlines are only `1px / resolution` *right now*, but won't necessarily be in the future with higher-res screens, still applies.

There's also probably a reasonable accessibility argument that UAs might want to offer the ability to force "thick" hairlines (maybe tied into other options that are about vision issues). So offering this ability at the low level of authors just supplying a tiny length makes that more difficult, but Kevin's idea makes it pretty trivial. Authors already *should* be allowing for the possibility of a 1px "hairline", given the continuing existence of 96dpi devices, but I give good odds on most authors assuming that hairlines are *necessarily* thinner than 1px because they look that way on all their devices.

**Comment by webstrand** [comment]
I've been fighting with some displays that have non-integer device-pixel-ratios, you can simulate this on most browsers by pressing `ctrl +` to zoom in. I do not need to address individual device pixels but I do need to ensure that my boxes are actually aligned to the pixel grid. The only solution I've found is to inject `--dpr` and use `round(whatever, var(--dpr) * 1px)`.

Here's an example that suffers badly from non-integer DPR: <https://codepen.io/webstrand/pen/OPLjrLy> notice that the gaps around the "check" of the checkbox are uneven. I do not mind if it's more than a 1px gap, it just needs to be even and consistent. Not this:
![image](https://github.com/user-attachments/assets/9ed720ec-ab9f-4297-b710-9a3ce32f35e0)
![image](https://github.com/user-attachments/assets/f21bc090-cf8e-497a-a049-e0af0416f245)

**Comment by webstrand** [comment]
Also notably, issues arise when dealing with hairline `gap: 1px` that would not be solved by adjusting `border-width` alone.

**Comment by tabatkins** [comment]
Summarizing the proposal again:

* Add a 'hairline' keyword to 'border-width' (and possibly 'column-rule-width'/'row-rule-width', does it make sense anywhere else?)
* This is equivalent to a 1px size, but *paints* differently - for borders it paints a "hairline" (likely device-pixel aligned) along the outer edge of the border region, with the rest of the border region being transparent (showing the background-color underneath, as normal).  (Equivalent to `border-color: stripes(transparent 1fr, currentcolor 1devicepixel)` or so.) For column rules it paints a hairline in the center of the 1px region.

Benefit of this is that the layout size is predictable (always 1px). It also doesn't require any special handling in other properties to keep things aligned; hairlines are 1px wide. And if a user has low-vision needs, a UA could offer to paint `hairline` thicker (up to `1px`), without potentially triggering layout issues in the page.

Downside of this is that it can still look odd for background *images* (they don't paint under borders, so you'll see a <1px stripe of background-color around the image). It also doesn't actually make the element smaller, in case that matters.

There's also a question of what to do for collapsed table borders. Reading Tables 3 and looking at Blink's behavior (and assuming, probably wrongly, that they match), I think the right answer is that the hairline should be center-aligned in this case. This is pretty obviously the only right answer for border *between* cells, and it lines up with the sizing and positioning of cell backgrounds (in Blink, at least), which only extend halfway into the border area. It means the borders won't be on the outside of the table grid, possibly revealing background bleed there, but, uh, backgrounds on tables seem pretty messed up anyway from what i can see.

-------

The alternative behavior suggested is to rely on border-width rounding, which always rounds away from 0:

* encourage authors to simply write a "small" border-width value, like `border-width: .1px`
* define [border rounding](https://drafts.csswg.org/css-values/#snap-a-length-as-a-border-width) to clamp to the "hairline" width, rather than to device pixels 
* add a `border-round(<length>)` function that exposes this rounding behavior directly, so other properties can lean on it and get the same size.

Benefit of this is that you get "real" sizing of the border area, and don't have to worry about it being transparent in some areas and potentially showing something confusing thru. You can also align other things to that size, if needed.

Downside is that it's an unpredictable size, which we know from experience can trip authors up when they don't test on different devices, tho to be fair it's only *slightly* unpredictable (smaller than the historical "px vs pt" sliding differences that browsers had to abandon). It also relies on author *guessing* at a border size which is small enough to round up to hairline, but not *so* small that it gets truncated to zero (especially a concern since browsers seem to all use some notion of "layout units" for sizing borders/etc, which are fixpoint and can't be smaller than .01px or so). I think it would be safe for the spec to recommend .1px for this, but it's still a little fraught. We could also still just add a 'hairline' keyword to represent this size directly, so no guessing involved.

-------

@emilio, you had more of an opinion towards the "use border-width rounding" option. Is that still your preference? Do you have other reasons, not listed, to prefer it?

@smfr, you also seemed to lean towards using border-width rounding. Is that accurate?

**Comment by smfr** [comment]
I am strongly opposed to the "paint a hairline but use a 1px width", for reasons discussed. I don't yet have an opinion on rounding.

**Comment by smfr** [comment]
And `outline-width`.

**Comment by Loirooriol** [comment]
Also sizing properties? I could imagine

```css
hr {
  height: hairline;
  background: linear-gradient(to right, transparent, #000, transparent);
  border: none;
}
```

**Comment by tabatkins** [comment]
> Also sizing properties?

I think that's a *much* smaller use-case, and I'd be fine with letting authors use `border-round()` for that. ^_^

**Comment by emilio** [comment]
@jfkthame that is already the case tho, if you have e.g. border-width: 0.5px on a hidpi screen and you move to a 1x display (or zoom in/out)

**Comment by tabatkins** [comment]
Okay, so new proposal:

* Add a `hairline` keyword to a few properties, which represents some fraction of a pixel that is still considered "visible", and it always ["snapped as a border width"](https://drafts.csswg.org/css-values-4/#snap-a-length-as-a-border-width) (so it's a positive integer number of device pixels, to the extent that makes sense).
* Introduce a `border-round()` function that takes a length (or the keyword `hairline`) and resolves to that length "snapped as a border width", so you can get this behavior for other properties that we don't define `hairline` for.

Properties that gain `hairline` are:
* 'border-width'
* 'column-rule-width'/'row-rule-width'
* 'outline'
* 'text-decoration-thickness'

Agenda+ to see if we can resolve on this.

**Comment by Loirooriol** [comment]
> represents some fraction of a pixel

Is this fraction left up to the UA to decide?

 - If not, e.g. we define that `hairline` means `border-round(0.1px)`, then it seems like we could just avoid adding `hairline` and let authors use `--hairline: border-round(0.1px)` if they want.
 - If yes, why only add `hairline` to a few properties? Why not make it a unit?

**Comment by edokeh** [comment]
On Safari, 0.1px or 0.2px borders may not render.

See this bug https://bugs.webkit.org/show_bug.cgi?id=286256

**Comment by SebastianZ** [comment]
As we consider to expose device pixel rounding as a function, I think it makes more sense to call it something related to that like `round-to-device-pixels()` (possibly something shorter) instead of `border-round()`.
Also, the algorithm to ["snap a length as a border width"](https://drafts.csswg.org/css-values-4/#snap-a-length-as-a-border-width) should be renamed accordingly.

Sebastian

**Comment by bernhardf-ro** [comment]
For both `hairline` and the rounding function (implicit or explicit) it would be great if output media with no device pixel resolution, e.g. PDF, would be covered by the specification as well.
The primary question is whether an infinite resolution should be used, making rounding pass through the input value unmodified, or a fixed resolution should be applied, e.g. `2*96dpi`, `4*96dpi`, `4*72dpi`, `300dpi`.

**Comment by webstrand** [comment]
@SebastianZ why not reuse the existing `round(up, 0.5px, 1dpx)` function? Then you can reuse the existing rounding modes provided by that function. Ultimately, if you expose a `device-pixel-round()` authors will be able to use it to derive a `1dpx` unit _anyway_. You could restrict the unit `1dpx` to be a singleton, so `2dpx` would be invalid (or extensible in the future), forcing authors to resort to `calc(100 * 1dpx)` if they actually want to use the unit. This is probably sufficiently discouraging to prevent casual use of the unit?

In my sites I currently inject a:
```css
@property --dpx {
	syntax: "<length>";
	inherits: true;
	initial-value: 1px;
}
:root { --dpx: 0.8666666666666667px }
```
via javascript so that I can make things look non-fuzzy. My displays have a native non-integer ratio to css pixels, which is why the size is fractional.

Used like `round(var(--thumbnail-width), var(--dpx))` it works very well in solving the problem in chrome/firefox.

**Comment by webstrand** [comment]
Device pixels may not be a reasonable for rendering `border-width`s, but if there's a smaller addressable unit than `1hairline` you get things like css grid distributing space in surprising ways (for example <https://codepen.io/webstrand/pen/MYYGqWQ>). 

So `1hairline` needs to be defined as "smallest layout unit that's an integer multiple of device pixels" so that layouts like flex/grid can't split `1hairline` in half and distribute its space between elements.

**Comment by Loirooriol** [comment]
@tabatkins You have explained the problems with having a "1 device pixel" unit.

But I think you haven't explained the reasons for not making `hairline` a `<length>` unit. https://github.com/w3c/csswg-drafts/issues/3720#issuecomment-2667147267

**Comment by tabatkins** [comment]
Yeah, I think a unit would also be acceptable, it's just slightly weirder. But I suppose it does work better as a unit when you're doing anything non-trivial, like a calc().

**Comment by smfr** [comment]
But what is `100hairline`?

**Comment by tabatkins** [comment]
100 times the hairline width? I'm not sure I understand the intent of your question.

**Comment by tabatkins** [comment]
`100hairline` might be weird, but `calc(100px + 2hairline)` (to match the size of another element with `width: 100px; border: 1hairline solid black;`) seems perfectly reasonable.

**Comment by kizu** [comment]
I wonder if it could be a keyword similar to `pi` or `e`?

And additionally allow it without `calc()` around it for `length`? (adding to the `<length>` syntax, I guess?)

Then, both `border: hairline solid;` will make sense, and you could use it in a `calc()` like `calc(100px + 2 * hairline)`?

**Comment by Crissov** [comment]
`hairline` is acceptable as a border width _keyword_ or environment _variable_, but feels icky as a _unit_, even if shortened to `hl`. I would suggest `dot` in that case, but I think enough arguments have been provided against this solution.

I had also considered constants like `pi` earlier today, but they are specifically numeric and their value must not depend on implementation or device. That is what `env()` is for – isn’t it?

**Comment by tabatkins** [comment]
So, here are the options:

1. Add a `hairline` keyword (so `border-width: hairline` works), and a `hairline` calc-keyword (so `width: calc(100px + 2 * hairline)` works).
2. Add a `hairline` unit (so `border-width: 1hairline` works, as does `width: calc(100px + 2hairline)`)
3. Add a function (so `border-width: border-round(0.1px)` works, as does `width: calc(100px + 2 * border-round(0.1px))`)

2 and 3 both have the advantage that they work everywhere, automatically, with no grammatical ambiguity. We don't have to review any existing properties, or keep this in mind with any future properties.

1 doesn't have that advantage. The plain keyword is grammatically ambiguous anywhere a length is mixed with an open set of keywords. We *mostly* nowadays make sure that open sets of keywords use dashed-idents, but it's still not 100%, and there are several legacy uses that might be ambiguous now. (I haven't surveyed, I'm going on intuition.) But it *is* certainly the simplest, best-looking of the options otherwise.

Some responses from the late minutes that I couldn't get to during the call:

> fantasai: i think making this a unit feels wrong, if you want the boxes to match you should use box-sizing

That only works *sometimes*, if you can make sure that the box you're matching uses a specifiable border-box size. The example I was giving doesn't allow that - the *content* box is specifiable (100px), plus two hairline borders, and so the matching box needs to be 100px + two hairlines.

> fantasai: people might start using the unit to reference device pixels in a way that's more fragile than we want

I don't see how any of the options are more or less likely to do this. In *all* of them, you can reference device pixels in some way. `border-round(.1px)` is exactly the same as `1hairline`, just slightly longer to spell, and gives exactly the same abilities.

A *very limited* solution that just allowed the keyword `hairline` (and *not* as a calc keyword, only usable by itself, kinda like an intrinsic size keyword) would indeed make it harder to misuse, but only by making it harder to use *in general*. And it wouldn't rule it out anyway; you could use multiple elements to achieve the same effects in many situations.

**Comment by SelenIT** [comment]
What about making it an `env()` var, as @Crissov suggested above? It seems to have all the advantages of the option 3 to me 🤔

**Comment by tabatkins** [comment]
An `env()` would work, yes - functions are grammatically unambiguous in any location. It's a little long (`env(hairline)`, presumably?), but not too bad.

**Comment by Loirooriol** [comment]
I'm still a bit confused about the difference with 1 device pixel.

Is the idea that https://drafts.csswg.org/css-values-4/#snap-a-length-as-a-border-width will then ensure that non-zero border widths are at least 1 hairline instead of the current 1 device pixel? For values above that, will the rounding be to an integer number of hairlines or still device pixels?

**Comment by tabatkins** [comment]
The use case is not actually "render a border 1 device pixel wide". If you have a thoeretical future screen with 100dppx, you don't want the border that small - it would be practically invisible. The use-case is "render a border thinner than 1px, if it's possible for the screen to do so while maintaining the sharp line we expect from borders". You can do lines that thin in print, and it's totally reasonable for many cases. So providing a `dp` unit would just mean that, in a future where screens that high-res exist, we'd have to redefine `dp` to be a "virtual device pixel", like we do with `px` today, and keep it from getting too small, or else sites would end up with effectively invisible UI that the author did not originally intend.

> Is the idea that [drafts.csswg.org/css-values-4#snap-a-length-as-a-border-width](https://drafts.csswg.org/css-values-4/#snap-a-length-as-a-border-width) will then ensure that non-zero border widths are at least 1 hairline instead of the current 1 device pixel? For values above that, will the rounding be to an integer number of hairlines or still device pixels?

Min of hairline, increments of device pixels.  (There's no guarantee that a hairline width will evenly divide 1px - on a 5dppx screen, a hairline of 2dp might be reasonable, for instance.)

**Comment by Loirooriol** [comment]
> increments of device pixels

Then this means that `border-round(x)` can't completely be expressed in terms of `env(hairline)`.

In your example, `border-round(0.7px)` would round down to 3dp (0.6px) , while `round(down, 0.75px, env(hairline))` would round down to 2dp (0.4px).

**Comment by benface** [comment]
> The use case is not actually "render a border 1 device pixel wide". If you have a thoeretical future screen with 100dppx, you don't want the border that small - it would be practically invisible. The use-case is "render a border thinner than 1px, if it's possible for the screen to do so while maintaining the sharp line we expect from borders".

If that's the use case, isn't it already possible with e.g. `0.5px`? I imagine the answer lies in "while maintaining the sharp line we expect from borders", but I don't see what's not sharp about it.

**Comment by benface** [comment]
Ah, I think I see the problem.

```css
div {
  --border-width: 0.5px;
  border: var(--border-width) solid black;
  width: calc(100px + 2 * var(--border-width))
}
```

Presumably, the intention is to make the content's width 100px, but on screens where the `0.5px` border-width is rounded to `1px` because the physical pixels are too large, it would actually make the content width 101px minus 2px, so 99px.

Is that correct? If so, then `border-round` or `round(..., ..., env(hairline))` make sense to me!

**Comment by benface** [comment]
Here's an idea, I don't know if it's a good one:

- `100hl` = `border-round(1px)` = `1px`
- `50hl` = `border-round(0.5px)` = `1px` on low density monitors, `0.5px` on the first Retina displays, etc.
- `1hl` = `border-round(0.01px)` = 1 physical pixel (technically could be more, but unlikely to ever be)

**Comment by emilio** [comment]
Ironically I think I just came across a decent use case for the naive env() approach while cleaning up the Firefox titlebar...

Consider an icon font. If you want crisp icons, you want the font-size snapped to real device pixels. That's trivialish to do with an hypothetical `env(resolution)` and `round()`.

Otherwise you end up with hacks like these: https://searchfox.org/mozilla-central/rev/beba5cde846f944c4d709e75cbe499d17af880a4/browser/themes/windows/browser.css#209 (those are for svgs but I'm switching them to text).

Not sure that's enough to justify it, and I might be missing something, but... :)

**Comment by emilio** [comment]
Well paging this discussion back I guess most proposed solutions would actually work with this use case one way or another, presuming that a hairline will always be a device pixel multiple. So nvm

**Comment by tabatkins** [comment]
Right, a hairline will be a multiple of device pixels anyway, since borders *need* to be device pixel-snapped. So `round(nearest, 1.2em, 1hairline)` (or however you'd spell the numeric value version of it) would work. It's not as theoretically finely-grained as it *could* be, but per the definition of a hairline, it's almost certainly more than sufficient, I think.

**Comment by tabatkins** [comment]
> In your example, border-round(0.7px) would round down to 3dp (0.6px) , while round(down, 0.7px, env(hairline)) would round down to 2dp (0.4px).

Sure, that's indeed the case. I don't think it's very important, tho?

**Comment by Loirooriol** [comment]
My point was that if `border-round()` can't be expressed in terms of 1 hairline, then we may need both features?

**Comment by GabbeV** [comment]
I found this issue after debugging a UI like this https://codepen.io/GabbeV/pen/EaVjeWw and thought it might be a good use case to consider. The borders are overlapped with a negative margin between items but because the values get rounded differently you get weird results when using some non integer dpi/zoom factors in your browser. Both border-round() and hairline could be used to solve the issue however border-round() would allow you to choose the size of the border instead of getting forced to use some multiple of hairline. This is not really an argument for or against introducing hairline but an argument for why border-round() would be useful regardless.

**Comment by emilio** [comment]
FWIW I prototyped `env(hairline)` in Firefox (https://hg.mozilla.org/mozilla-central/rev/f939c46a47f1) and so far I found it pretty painless to use.

**Comment by emilio** [comment]
Another use case that's hard to do without this or `border-round()` is `outline-width: some-value; outline-offset: -some-value;` (as they won't match in fractional DPIs)

**Comment by tabatkins** [comment]
Yeah, Agenda+ to just resolve on `env(hairline)`.

**Comment by tabatkins** [comment]
@Loirooriol 
> My point was that if border-round() can't be expressed in terms of 1 hairline, then we may need both features?

*My* point was asking whether it *matters*, tho. ^_^ The difference here is *less than the width of a hairline* - in what cases is that a meaningful difference that authors will care about, even when the DPI is high enough that there's technically device-pixel lengths between the hairline widths?

**Comment by webstrand** [comment]
If the two different measures are directly adjacent, then it'll be visible even at high DPI, not because the individual pixels can be seen but because of the offset in light intensity. The visibility of less-than-hairline lines is contrast dependent.

**Comment by tabatkins** [comment]
I think I'm happy waiting for such high-DPI screens to exist and start causing enough problems for someone to complain, rather than trying to pre-emptively solve the issue without ever experiencing it. ^_^

**Comment by Merci-chao** [comment]
Sorry for spamming things that someone may have mentioned above already. There are too many replies above, thus tl;dr...

According to the definition:
https://drafts.csswg.org/css-values-4/#snap-a-length-as-a-border-width
> To snap a length as a border width given a <length> len:
> 1. Assert: len is non-negative.
> 2. If len is an integer number of device pixels, do nothing.
> 3. If len is greater than zero, but less than 1 device pixel, round len up to 1 device pixel.
> 4. If len is greater than 1 device pixel, round it down to the nearest integer number of device pixels.

Here's a not quite pure CSS polyfill for `border-width()`:
```css
--device-pixel-ratio: 1.25;
--css-width: 2px;
--border-width: calc(
	max( /*at least 1 device px, or 0 if size <= 0*/
		round( /*round down to integer*/
			down,
			var(--css-width) * var(--device-pixel-ratio), /*covert to device px*/
			1px
		),
		min( /*will only be 1px (size > 0), or NaN (size <= 0) which forces the final result to be 0px*/
			max(var(--css-width), 0px) / 0, /*will only be Infinity or NaN*/
			1px
		)
	) / var(--device-pixel-ratio) /*covert back to css px*/
);
```

Here's the demo that trying to display an inset outline to overlap the border (as [emilio](https://github.com/emilio) mentioned [above](https://github.com/w3c/csswg-drafts/issues/3720#issuecomment-3224951042)):
[demo.htm](https://github.com/user-attachments/files/22712920/demo.htm)
```html
<body>
<div id="a">A</div>
<div id="b">B</div>
<style>
div {
	--css-width: 1px;
	--border-width: calc(
		max( /*at least 1 device px, or 0 if size <= 0*/
			round( /*round down to integer*/
				down,
				var(--css-width) * var(--device-pixel-ratio), /*covert to device px*/
				1px
			),
			min( /*will only be 1px (size > 0), or NaN (size <= 0) which forces the final result to be 0px*/
				max(var(--css-width), 0px) / 0, /*will only be Infinity or NaN*/
				1px
			)
		) / var(--device-pixel-ratio) /*covert back to css px*/
	);
	width: 100px;
	height: 100px;
	border: var(--css-width) solid red;
	outline: var(--css-width) dashed lime;
	margin: 1em;
}
#a { outline-offset: calc(var(--css-width) * -1) }
#b { outline-offset: calc(var(--border-width) * -1) }
</style>
<script>
observeDPRChange(() => document.body.style.setProperty("--device-pixel-ratio", devicePixelRatio));

function observeDPRChange(callback) {
	matchMedia(`(resolution: ${devicePixelRatio}dppx)`)
			.addEventListener("change", e => observeDPRChange(callback), {once: true});
	callback();
}
</script>
</body>
```
Box A applies the traditional method, but usually fails on fractional DPR; Box B applies the polyfill above and has a consistent appearance on any DPR.

Don't understand why it is not worthy to implement the `border-width()` (or any other name). It is very annoying for most web developers that the rendered border width is not predictable and layout may fail unexpectedly.

**Comment by emilio** [comment]
I don't think it's so complicated tho, wouldn't:

```
--hairline: calc(1px / var(--device-pixel-ratio));
--border-width: max(var(--hairline), round(var(--css-width), var(--hairline)));
```

work for that use-case? Again not opposed to add a more general `border-round` for convenience, but I think `env(hairline)` unlocks most of the use-cases already, and we can resolve on `border-round` on a separate issue :)

**Comment by Merci-chao** [comment]
@emilio 
The logic I posted is complicated because it includes these 2 rules in the definition:
>  Assert: len is non-negative.
>  If len is greater than zero,

If we omit the control logic for these two assertions:
```css
min( /*will only be 1px (size > 0), or NaN (size <= 0) which forces the final result to be 0px*/
	max(var(--css-width), 0px) / 0, /*will only be Infinity or NaN*/
	1px
)
```
And all the comments, the code will be:
```css
--border-width: calc(
	max(
		round(
			down,
			var(--css-width) * var(--device-pixel-ratio),
			1px
		),
		1px
	)
	/ var(--device-pixel-ratio)
);
```
And here is the code in a single line:

```css
--border-width: calc(max(round(down, var(--css-width) * var(--device-pixel-ratio), 1px), 1px) / var(--device-pixel-ratio));
```

It uses a total of 2 functions (not counting `calc()`), 3 variables, 2 operators, and 2 constants.


Your version looks simpler because it defines two variables. If you change it to use only one variable, then:

```css
--border-width: max(1px / var(--device-pixel-ratio), round(var(--css-width), 1px / var(--device-pixel-ratio)));
```

It uses a total of 2 functions, 3 variables, 2 operators, and 2 constants. Isn't it just as complicated? Well, I do agree that using `--hairline` for reusability is a good idea.

As a reminder, it needs to be rounded `down` but not `nearest`:
> 4. round it down to the nearest integer number of device pixels.

Your code breaks, for example, when the zoom level is set to 290% (`--device-pixel-ratio: 2.857142857142857`) because of that. Also, be careful that it doesn't support `--css-width: 0px`, as that will result in 1 hairline, which is incorrect.
[demo.htm](https://github.com/user-attachments/files/22722912/demo.htm)

<img width="765" height="677" alt="Image" src="https://github.com/user-attachments/assets/63d0d308-e3eb-4fab-ad76-1a1de197843d" />


===============

P.S. 
I tried modifying your version to use `round(down)`, but it breaks when DPR is 3. This suggests there may be a fundamental flaw in the logic of your implementation.

[demo.htm](https://github.com/user-attachments/files/22723451/demo.htm)

<img width="1000" height="898" alt="Image" src="https://github.com/user-attachments/assets/919b8f5f-b660-4d4f-b3bd-fc024fc3b5ea" />

**Comment by Loirooriol** [comment]
The proposal was to add `env(hairline)`, not `env(device-pixel-ratio)`. Most of the complexity in your code comes from that.

But if you want to replicate the "snap a length as a border-width" exactly, then [you can't do it with just `env(hairline)`](https://github.com/w3c/csswg-drafts/issues/3720#issuecomment-2898984858).

So I'm not sure I get your point. Is it that `env(hairline)` doesn't suit your needs and you would like `border-round()` too?

**Comment by Merci-chao** [comment]
> The proposal was to add `env(hairline)`, not `env(device-pixel-ratio)`. Most of the complexity in your code comes from that.
> 
> But if you want to replicate the "snap a length as a border-width" exactly, then [you can't do it with just `env(hairline)`](https://github.com/w3c/csswg-drafts/issues/3720#issuecomment-2898984858).
> 
> So I'm not sure I get your point. Is it that `env(hairline)` doesn't suit your needs and you would like `border-round()` too?

Uh… there are way too many replies above. I only skimmed a few, so I’m not really sure what everyone’s asking for.

I'm talking about a function `border-width()` (or similar), to, say, make an outline perfectly overlap the border in any device and any DPR:

```css
--width: 1px;
border: var(--width) solid red;
outline: var(--width) dashed lime;
outline-offset: calc(border-width(var(--width)) * -1);
```

If `env(hairline)` works similar, e.g.:

```css
--width: 1;
border: calc(1px * var(--width)) solid red;
outline: calc(1px * var(--width)) dashed lime;
outline-offset: calc(env(hairline) * var(--width) * -1);
```

So, uh… isn’t that basically the same thing? Just a different way of doing it.


=======

P.S. Sorry, I’m a bit confused of what exactly is `env(hairline)`  for...

**Comment by Merci-chao** [comment]
If you’re trying to, say, size a box to 100dpx by 100dpx, with a 15dpx width border, I mean, isn’t that how it works?

[demo.htm](https://github.com/user-attachments/files/22725812/demo.htm)

I've even tried it on my phone, Firefox Android, with using DPR = 3, it does show as 100dpx * 100dpx, 15dpx border.

I'm confused...

<img width="1080" height="483" alt="Image" src="https://github.com/user-attachments/assets/7508df16-d141-4de7-a081-a2447d4b02ba" />

**Comment by Merci-chao** [comment]
I tested it on Chrome Android, and [`demo.htm`](https://github.com/user-attachments/files/22725812/demo.htm) works as expected too.

<img width="1080" height="595" alt="Image" src="https://github.com/user-attachments/assets/e960df07-084c-4d59-ada1-afe5d55ea85f" />
 
So now I’m actually the one unsure what the point is...

> The proposal was to add `env(hairline)`, not `env(device-pixel-ratio)`. Most of the complexity in your code comes from that.

Back to your question. I’d say it’s better to add `env(device-pixel-ratio)` instead of `env(hairline)` (I’m still a bit confused about what “hairline” means. Is it just 1 device pixel?). The reason is: you can use DPR to convert between CSS pixels and device pixels, but you can’t use a CSS pixel value and a hairline (or device pixel?) value to reverse-engineer the DPR for other purposes. As far as I know, CSS doesn’t currently support multiplication or division between values with units.

If folks really hate any complexity, why not propose both `env(hairline)` and `env(dpr)`?

Please don't blame me for the complexity in the `border-width()` logic I posted (or `border-snap()` I’d recommend). It's not me, it's just that border snapping itself is complex:
https://drafts.csswg.org/css-values-4/#snap-a-length-as-a-border-width
> To snap a length as a border width given a len:
> 1. Assert: len is non-negative.
> 2. If len is an integer number of device pixels, do nothing.
> 3. If len is greater than zero, but less than 1 device pixel, round len up to 1 device pixel.
> 4. If len is greater than 1 device pixel, round it down to the nearest integer number of device pixels.

CSS doesn't support if-else logic to handle the assertions, and thankfully there are some hacks that can work around that.

In the end, I still don't get why folks are so allergic to complexity and want to strip it all away. I'm all for simplification too, but not at the cost of ignoring details and just hoping everything magically works.

**Comment by emilio** [comment]
I think the negative outline offset use-case is probably worth addressing separately fwiw, I filed #12906 about that.

**Comment by webstrand** [comment]
The negative outline offset use-case is going to be hindered by CSS half-pixels. If the containing element is not already aligned to the device-pixel grid, then it'll have uneven gaps between child elements, preventing the negative offset from correctly overlapping as desired. You can simulate this for yourself by overriding your display to have a non-integer DPR, or on some browsers using the "zoom" feature also causes non-integer DPR rendering for the page.

I'd be in favor of `env(device-pixel-ratio)`, you can already inject this with javascript, and you can use it to solve both the box alignment and border-width issues.

I know previously it's been expressed that `device-pixel-ratio` is meaningless in print, but the same is true for `env(hairline)` or `border-round()` - print doesn't have the same sub-pixel rendering constraints as screens. For print contexts, `env(device-pixel-ratio)` could simply return 1 (or the printer's actual resolution ratio if available), just as hairline borders would need some sensible fallback behavior for print media.

**Comment by emilio** [comment]
> The negative outline offset use-case is going to be hindered by CSS half-pixels.

What do you mean? If the browser snaps the border / outline to the device-pixel grid, rounding the outline to device pixels should also snap them to the pixel grid, what am I missing?

**Comment by robizzt** [comment]
Also had the same problem: I set the scaling factor on my phone, which resulted in a DPR of 2.625, so the border width was 0.762px instead of 1px.

This works for me:
```
--dpr: 1; /* Device Pixel Ratio - Will be overridden by JS */
--hairline: round(down, 1px, calc(1px / var(--dpr)));
```

Then use it on any other CSS property:
`padding-bottom: var(--hairline);`
or
`padding-bottom: calc(var(--hairline) * 2);`

**Comment by fantasai** [comment]
There's been a lot of use cases brought up here, and I'm not sure that they all are converging on the same solution:

- Drawing a hairline border, which is at least one device pixel but may be more. Reasonably handled via a new keyword for `border-width` as proposed in the OP.
- Insetting outlines. Probably better handled via https://github.com/w3c/csswg-drafts/issues/12906
- Calculating box sizes to include pixel-snapped border-widths. Likely handled better via `box-sizing` or https://github.com/w3c/csswg-drafts/issues/7770
- Overlapping borders with negative margins. See [this comment](https://github.com/w3c/csswg-drafts/issues/3720#issuecomment-3075180734). Requires either a border-rounding function or the ability to reference the used `border-width` as a `margin` keyword value.
- Snapping sizes of things (such as images) to device pixels. See [this comment](https://github.com/w3c/csswg-drafts/issues/3720#issuecomment-2989276604) from Emilio.  Probably best handled by adding a `pixel` argument to the `round()` function, but also brings up questions about pixel-snapping the position of the box.

**Comment by tabatkins** [comment]
All right, disappointingly, we didn't come to resolution today, because `env(hairline)` was blocked on us first solving the more general problem of actual device-pixel lengths and/or border-rounding. [My Agenda+ request](https://github.com/w3c/csswg-drafts/issues/3720#issuecomment-3226053230) was specifically trying to push those more general problems to the side *so they stop blocking this issue from 2019, which has been around since many years earlier*.

I still don't understand the logic behind this blocking. Assuming that someone comes up with the perfect dppx solution tomorrow, the use-cases for "I want this non-border property to be the same size as the hairline" still exists, so we'll still want to expose env(hairline). The use-case is identical for "I want this non-border property to be the same size as my `medium` border", but `medium` is just a synonym for `3px`, so that's completely doable; `hairline` doesn't have an exact synonymous size, by the nature of its definition.

The big objection from @fantasai, as far as I understood it, is the possibility of people misusing `env(hairline)` for things where they "really want" a device pixel length. Is this correct, @fantasai?

If so, both I and @emilio argued in the discussion that this isn't a meaningful problem in practice, and in fact it will often be *avoiding* a footgun. Most use-cases where people *think* they want device pixels, they actually just want "smaller than 1px and device-pixel-snapped", which `env(hairline)` provides (so long as the width is already flush with a dp boundary, of course, but that's no different than 1px only being crisp at a dp boundary; it usually will be). In many cases, if device pixels were suddenly 1/10th the size, their use-case would *break* if we actually gave them that size.

Even in cases where people *legitimately* do want device pixels, using a hairline that might be a few dp in size is almost never *bad*, it's just *not quite optimal*. That's not, imo, enough of a problem to object to the length, given the practical use-cases presented.

Device pixels are a *very* dangerous thing to actually expose to people. Authors have ways to obtain that length if they really want, via a little bit of scripting, but making it easy to obtain in plain CSS is dangerous. As we move forward I think it's likely we *will* expose more dp details to authors, but in use-case-specific ways, where it's harder to misuse.


-------

So, in conclusion:

* "Real" device pixels continue to be a dangerous thing to expose, and most people asking for it don't actually want it and would have their pages break if we gave it to them.
* People using `env(hairline)` when they "really want" device pixels is totally fine in practice. Using it is often *better* than using "real" device pixels, and in the rare cases where their use-case is actually reasonable to solve with real device pixels, using `env(hairline)` instead doesn't *break* anything, it just isn't *quite as optimal* as real device pixels would be (on theoretical future ultra-high-res screens).
* Use-cases for "real" device pixels are extremely varied, and often pull in questions of dp-snapping of geometry, of responding to zoom, of printing and variable color vs black resolution, etc. There is **no generic solution** that solves all the dp-related use-cases.
* There are many practical use-cases for a hairline width (or something very similar, per previous point) outside of borders and rules, so just adding a `hairline` border-size keyword that computes to a <=1px length isn't sufficient.

So, my requested resolution next time is that we affirmatively push dp-related shenanigans to a separate issue, and resolve on adding `hairline` to `<line-width>` in the Borders spec and `env(hairline)` to the Env spec.

**Comment by webstrand** [comment]
I don't fully understand the danger of device-pixels, is the thrust of the argument: if we let people draw lines at 1dppx or 2dppx, it may look fine on their 96dpi display, but at 300dpi those features are now invisible?

I unfortunately have to deal a lot with displays that have a _native_ non-integer devicePixelRatio that are close to but not actually 96dpi. This means that not only are 1dppx alignment errors very obviously visible, but any user that uses the zoom function of their browser immediately breaks any layout that uses 1px boundaries, like the aforementioned `gap: 1px` because there's no other good way to draw border-collapsed style borders on top of a css grid. And due to not being snapped, this results in some gaps being anywhere from 0px to 2px wide.

Nearly all of my issues are solved by using `round(1px, var(--dpx))`.

The only other case where I've drawn something at 1dppx was a striped background gradient, where the goal was to have almost imperceptible striping. This was done with `max(1dppx, 0.5px)` because the effect looks better on high-dpi displays but needed to not become solid color on low-dpi.

---

I could see an argument where instead of giving access directly to dppx, you defined a hairline unit that was guaranteed to be an integer multiple of device pixels. To avoid the disappearance of features on high-dpi displays. But you'd need to modify layout so that boxes can't be laid out to sub-hairline alignments?

There would need to be an escape hatch, to allow users to layout boxes aligned with colored regions in images, which may have sub-hairline alignment of colored regions. For instance aligning a button on top of a background image, would look strange if :hover styles caused it to be 1dppx offset from the background due to hairline layout constraints.

So a hairline unit may not actually be a solution.

---

Print is definitely a concern as are non-rectilinear displays. The former has such high dpi that my issues with non-integer devicePixelRatio is moot: even if the pixels are fractional the printer has enough resolution beneath that to avoid visibly misaligning boxes. The latter I admit I don't comprehend the issue. The software canvas will be rectilinear pixels, but when rendered to the screen it will be mosaiced by software to an appropriate pattern for the display?

---


Without device pixels or hairline-aligned layout this always results in bad looking rendering when the device pixel ratio is non-integer.

```html
<style>
box {
    border: 1px solid black;
    background: black content-box;
    padding: 1px;
}
</style>
<box>x</box>
```

<img width="31" height="39" alt="Image" src="https://github.com/user-attachments/assets/f6de4e7e-9d8a-4b8a-b588-b4dacf4d286d" />

having access to rounding to device pixels may not solve every issue easily, but it does make them possible to solve. Grid layout gaps when the grid box itself is a non-integer number of device pixels wide can still only be solved with tedious `width: calc(math)` or using a container query to obtain inline-size and rounding down to the nearest multiple of dppx compatible with interior layout divisions. And yes I can use a short snippet of javascript to inject the `devicePixelRatio` as a custom property into the page, but I like to author css in such a way that it doesn't rely on the presence of javascript to function.

If you want to reduce the danger, expose the dppx keyword as a rounding mode only. It would mean that to round to multiples of dppx authors would need to do `round(nearest, 50px, round(up, 0px, dppx)`. Yes authors _can_ define their own `--dpx: round(up, 0px, dppx)` but that's opt-in and less of a foot-gun for the uninitiated?

---

possible final concern is how `dppx` or even `hairline` should behave in `scale: 0.5` or `zoom: 0.5` elements. Right now it is not possible for javascript to inject `--dpx` for individually scaled or zoomed elements, if its desirable to make the unit invariant to zoom. I'll note that it may actually be desirable to make it invariant to zoom, because some applications use zoomed iframes to simulate other display types inline. If dppx or hairline scales with the zoom, then the simulation is less accurate.

**Comment by tabatkins** [comment]
> is the thrust of the argument: if we let people draw lines at 1dppx or 2dppx, it may look fine on their 96dpi display, but at 300dpi those features are now invisible?

That's exactly it, yes. We have many examples across the years, both in CSS itself and in similar UI frameworks, where giving people access to a *variable-sized unit* that will have a stable size on *their* machines but not across their users' machines, or across time as hardware improves, leads to bad results.

People, very reasonably, use the tools that they have, in ways that appear to work for them at that moment. If they need a sharp, <1px line, and we give them device pixels, they'll try it out, see that a 1dp line looks good, and stick with it. Doing the "correct" thing of setting a *small but relatively static* line size and just *rounding* it to device pixels is more difficult, both to remember to do and simply to write, so most people won't do it.  There's no shame or shade in this; it's human nature, and it's something every language designer needs to keep in mind.

The "correct" thing to do wrt device pixels varies by use-case, which is why I think attempts to address them generically are doomed to fail. As you outline in the use-cases in your comment, there are a number of different possible concerns people might need to deal with when solving problems in the device-pixel space.

> I could see an argument where instead of giving access directly to dppx, you defined a hairline unit that was guaranteed to be an integer multiple of device pixels.

Yup, that's exactly the proposal. The `env(hairline)` value would be a browser-determined length that is just visible, rounded to the nearest integer number of device pixels (with a min of 1dp) at default zoom levels. `border-width: hairline` would be a slightly shorter way to specify the same thing, akin to `border-width: medium` being equivalent to `3px`.

> For instance aligning a button on top of a background image, would look strange if :hover styles caused it to be 1dppx offset from the background due to hairline layout constraints.

This sort of thing is already an issue authors have to deal with, and the common solution of using a same-width transparent border for the "no border" style would work just the same here.

**Comment by webstrand** [comment]
> Doing the "correct" thing of setting a small but relatively static line size and just rounding it to device pixels

Why not just make that the _only_ way to do it, like `round(down, 1px, dppx)`, introducing a keyword into the `round()` syntax? It does leave a backdoor for deriving device pixels, but javascript also provides a means. 

> The env(hairline) value would be a browser-determined length that is just visible, rounded to the nearest integer number of device pixels (with a min of 1dp) at default zoom levels.

I was proposing that if a hairline unit is provided, that **all layout** be quantized to that. If layout isn't quantized, then while it solves some of the issues, it doesn't fix issues caused by layout not being integer multiples of the fundamental unit.

For instance I mentioned the issues putting hairline borders around/between grid cells. If the hairline is larger than the layout unit, then you get uneven gapping (or borders that are too thick). 

<img width="200" height="70" alt="Image" src="https://github.com/user-attachments/assets/9408fd46-a574-4ec6-8486-ea2c18210573" />

<https://codepen.io/webstrand/pen/RNRpWzz>

The issue here is that the space the grid occupies is not a multiple of the device pixel, so after grid layout, there's an extra device-pixel left over. If this doesn't work for you, you most likely have an integer `devicePixelRatio`, try using `ctrl +` or `ctrl -` to zoom, to get a non-integer device-pixel ratio. Or find a non-integer ratio'd display, maybe a Apple Cinema HD?

On higher DPI screens like printer or Retina, the gap is less obviously an issue.

---

I don't care too much about whether the layout unit is a device-pixel or a hairline. Just so long as that's what everything gets quantized to prior to the paint phase.

**Comment by GabbeV** [comment]
I brought up `border-round()` in this thread originally as I believe it solves most of the issues discussed here.

The algorithm for rounding borders is already meant to ensure small borders remain visible. If its current exact definition doesn't do that for print I would suggest tweaking that definition. The same is true for potential future extreme resolutions. If clamping to 1 device pixel is too small on some future device then the device maker will likely ignore the spec and clamp to 2 device pixels to ensure the web works well on that device and the spec will eventually have to be tweaked to match that reality.

This makes me wonder why `hairline` is needed at all if you can do `--hairline: border-round(0.01px);` yourself and define something similar while retaining artistic control over the resolution required for it to become multiple actual device pixels.

The issues around overlapping borders with negative margins, outline offsets not rounding like borders or sizing a box to the same size as another that has borders are all solvable with `border-round()`. `border-round()` therefor seems like the missing primitive to allow you to solve all these issues.

`border-round()` does not have any of the issues that introducing a device pixel unit would have as borders are already rounded this way. How that rounding is handled when you zoom for example is already a solved problem. Both hairline and a device pixel unit would produce uncertainty around how to handle zoom. Do they increase in size or do they ignore the zoom? If you want to use them as a length I think increasing in size is the right answer but if you want to use them as a rounding interval in round() you probably want them to ignore the zoom, at least if you want to be able to emulate `border-round()` with them.

I originally didn't express any reservations against introducing `hairline`, even though I don't find it useful myself, as I see no issue with having it as well. However `border-round()` is in my opinion the missing primitive that actually unlocks things that aren't possible otherwise.

**Comment by Loirooriol** [comment]
> This makes me wonder why hairline is needed at all if you can do `--hairline: border-round(0.01px);`

Using 0.01px is kind of fragile, though. For example, a `0.008px` border doesn't render in Firefox, because before pixel snapping, it uses a unit precision of 1px/60, and 0.008 is closer to 0 than to 1/60. But Chrome seems to store it as a float before snapping, because even a `0.00000000000000000000000001px` border ends up being visible.

Note the 1px/60 precision is UA-specific, e.g. Blink uses 1px/64, and IE used 1px/100. So it's plausible that some browser would use a unit precision of 1px/30, and store the value using units (rounding to closest). Then `border-round(0.01px)` would be 0.

Having a hairline you wouldn't need to worry about this kind of problems.

**Comment by webstrand** [comment]
Are you imagining that `border-round(expr)` is semantically equivalent to `max(hairline, round(down, expr, 1dppx))`? My reservation is that by including "border" in the naming it means that any future semantics that get added to painting borders, might inappropriately get conflated here too.

I wouldn't want to inherit disappearing borders from Firefox when the value becomes too small, even if that's correct for borders.

**Comment by GabbeV** [comment]
I find it unlikely that you would ever actually need to use such high precision as the pixel density of the device would have to be quite ridiculous for it to matter. Even 0.01px is exaggerated. A more reasonable hairline would probably not be smaller than `--hairline: border-round(0.1px);`

**Comment by Loirooriol** [comment]
@webstrand I'm imagining `border-round(<length>)` will return the computed value that you would get with `border-top-width: <length>`.

@GabbeV Authors use exaggerated numbers, I have seen `z-index: 99999999999` lots of times. Lacking a hairline, someone might use `border-round(0.001px)` because it looks good in Chrome. But it won't work in Firefox.

**Comment by GabbeV** [comment]
@webstrand I'm imagining border-round being defined to do exactly the same thing as the algorithm for rounding border-width values. One way of getting there would be to make that algorithm expressible in CSS with `max()` and `round()`, however I think it would be better to have a dedicated function for it to avoid locking down that algorithm from ever changing. Making it expressible with `max()` and `round()` would also be very hard as it would require a device pixel unit that ignores zoom which I think would be very controversial to introduce. Maybe introducing a rounding interval keyword for device pixel would be more possible.

@Loirooriol I'm not against introducing `hairline` in addition to `border-round()` as I've stated. My argument is just that it is more of an arbitrary constant similar to `medium` that might be nice to have for some people but doesn't really solve the fundamental issues people are having in this thread.

**Comment by emilio** [comment]
I think the ideal solution would be to do all of:

 * A `hairline` keyword for line-widths. That's just convenience. I don't think anybody has any problems with this.
 * env(hairline) for use on everything. I think this can solve a lot of the issues that authors have.
 * A keyword to the `round` function (ideally) or a `border-round` function. I think adding an extra `line-width` keyword which would disallow a third argument to round() is probably slightly nicer? Anyways bikesheddable but it seems a requirement if we don't want to expose the plain dpx value, as otherwise you can't implement the right rounding behavior unless you assume env(hairline) is a device pixel.

I think we could do something parallel to 2 and 3 for device pixels, but separate issue :)

**Comment by frivoal** [comment]
> I still don't understand the logic behind this blocking.

As far as I am concerned, I am not strictly blocking `env(hairline)` until pixel snapping and such is solved. But I am saying we should take things one step at a time. Having a `hairline` keyword for `border-width` and the like does not solve all the use cases, but in the cases it does solve, it allows authors to express a clear intent, does exactly the right thing, and nothing more. It is both useful and safe, so I think we should do it immediately, and look how far that gets us.

Specifically, I think we should:
* add `hairline` to `<line-width>` (which is already used in `border`, `outline`, `gap-rule`)
* extend the syntax to allow `<line-width>` in a few of properties which arguably should have had it all along, but never bothered to go beyond `<length>` because the other values of `line-width` weren't thus far very useful: `padding`, `margin`, `text-decoration-thickness`, `gap`, `box-shadow-spread`, `stroke-size`

I agree that `env(hairline)` can be used to do useful things, and is a more generic tool. But I see that as a double edged sword:
* On the plus side, it can be used in places where we have not (yet) adjusted the grammar to allow a direct `hairline` keyword as part of the syntax's property, so people can do hairliny things without having to wait for the WG to accommodate their needs one by one, which relieves the WG of having to accommodate their needs one by one.
* On the flip side, because it's generic, it can be used in places where it doesn't really make sense. The problem with giving people a tool that is generically applicable, but does not necessarily make sense in a general context is that since it does something, people will use it and get something out of it, but the something may not be well behaved, and we get sites that work in some context, but are user hostile in others, work great on some device and poorly on others… This is especially true given that it can be used in `calc()`, which makes it effectively equivalent to a unit, and so people can use it not just for a hairline, but also to make something 300 hairlines tall, or giving things an aspect ratio of 1hairline to 1px, or whatever. Granted, sizing things in hairlines is not as brittle than sizing things in dpx. But it isn't a _generally_ sensible thing, so it is not _generally_ robust.

My proposal is not that we therefore never add `env(hairline)`. Rather, it is that we start with adding `hairline` to `<line-width>`, and using `<line-width>` a little more broadly as discussed above, and then wait to see what people do with it. My expectation is that:
* this already solves a substantial part of the problem space, especially if we give time to people to think creatively about how this interacts with the rest of the css toolbox
* this lets us separately look at the rest of the unsolved use cases, and possibly come up with additional solutions that address some subclass of them well without having weird side effects.

It is very possible that having done that, we will still conclude that `env(hairline)` is the best compromise of a tool that solves most problems without too many side effects. But I am not convinced of that, and I suspect that if we first pursue adding `hairline` to `<line-width>` and using `<line-width>` a little more broadly, we will find that we have solved more than we thought, and with a smaller reminder left to solve, we may be able to identify better solutions for that.

I am open to being proven wrong, and to the possibility that `env(hairline)` is indeed as good as its going to get. But my request is that we first do `hairline` with `<line-width>`, while we talk about `env(hairline)` and what else we might do to solve the rest of the solvable use cases.

**Comment by tabatkins** [comment]
Okay, I understand your concerns, Florian, but seem to disagree on the harm that authors could actually cause themselves. As you say, a hairline width is less likely to be a problem than a dppx width. People *can* do stupid things, but it's not meaningfully different from manually specifying a `.1px` or something, which they can do today (and don't).

So my preference is to accept [Emilio's "ideal proposal"](https://github.com/w3c/csswg-drafts/issues/3720#issuecomment-3762832472), that is:

* add `hairline` to `<line-width>`
* add `env(hairline)` to give the same width but generically
* add `border` as a rounding method to `round()` to solve some of the "just act like a border" use-cases that aren't necessarily hairline-related; that is, `round(border, Xpx)`

**Comment by fantasai** [comment]
We would support adding `hairline` to `<line-width>` and extending `<line-width>` to additional contexts (such as `gap`). We would also support adding the ability to round lengths the same way as `border-width` does. But we don't support adding `env(hairline)` at the moment, and would prefer to address the use cases for it more directly (for example as we did in https://github.com/w3c/csswg-drafts/issues/12906), because we don't want authors adding `round(..., env(hairline))` everywhere because they decided that snapping everything to device pixels was the clever thing to do.

In other words, we support @frivoal's proposal here.

**Comment by fantasai** [comment]
We also need to follow up with additional issues to address other use cases that were brought up, e.g.
- Should `gap` and all other properties that accept `<line-width>` snap like `border-width`?
- Should `line-height` snap to device pixels? Or `font-size`?
- Other use cases?

Also maybe one to decide if we want `round(line-width, ...)` or just `line-width(...)`.

**Comment by tabatkins** [comment]
All right, added 'hairline' keyword and `round(line-width, ...)`. Also added `<line-width>` to the definitions of gaps, text-deco-thickness, and stroke-width; @fantasai if there are any others you can think of please add as well.

> Also maybe one to decide if we want round(line-width, ...) or just line-width(...).

I quite like having this rounding functionality just live in `round()`; otherwise I have to duplicate more text. It looks pretty reasonable now that I've written it out.
